docs/rfcs/component-universal-telemetry.md

```diff
@@ -0,0 +1,119 @@
+
+### Attributes
+
+All signals should use the following attributes:
+
+#### Receivers
+
+- `otel.component.kind`: `receiver`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs`, `metrics`, `traces`
+
+#### Processors
+
+- `otel.component.kind`: `processor`
+- `otel.component.id`: The component ID
+- `otel.pipeline.id`: The pipeline ID
+- `otel.signal`: `logs`, `metrics`, `traces`
+
+#### Exporters
+
+- `otel.component.kind`: `exporter`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs`, `metrics` `traces`
+
+#### Connectors
+
+- `otel.component.kind`: `connector`
+- `otel.component.id`: The component ID
+- `otel.signal`: `logs`, `metrics` `traces`
+- `otel.output.signal`: `logs`, `metrics` `traces`
+
+Note: The `otel.signal`, `otel.output.signal`, or `otel.pipeline.id` attributes may be omitted if the corresponding component instances are unified by the component implementation. For example, the `otlp` receiver is a singleton, so its telemetry is not specific to a signal. Similarly, the `memory_limiter` processor is a singleton, so its telemetry is not specific to a pipeline.
+
+### Metrics
+
+There are two straightforward measurements that can be made on any pdata:
+
+1. A count of "items" (spans, data points, or log records). These are low cost but broadly useful, so they should be enabled by default.
+2. A measure of size, based on [ProtoMarshaler.Sizer()](https://github.com/open-telemetry/opentelemetry-collector/blob/9907ba50df0d5853c34d2962cf21da42e15a560d/pdata/ptrace/pb.go#L11). These are high cost to compute, so by default they should be disabled (and not calculated).
+
+The location of these measurements can be described in terms of whether the data is "incoming" or "outgoing", from the perspective of the component to which the telemetry is ascribed.
+
+1. Incoming measurements are attributed to the component which is _consuming_ the data.
+2. Outgoing measurements are attributed to the component which is _producing_ the data.
+
+For both metrics, an `outcome` attribute with possible values `success` and `failure` should be automatically recorded, corresponding to whether or not the corresponding function call returned an error. Specifically, incoming measurements will be recorded with `outcome` as `failure` when a call from the previous component the `ConsumeX` function returns an error, and `success` otherwise. Likewise, outgoing measurements will be recorded with `outcome` as `failure` when a call to the next consumer's `ConsumeX` function returns an error, and `success` otherwise.
+
+```yaml
+    otelcol_component_incoming_items:
```

Comment:
There is an advantage to have the `otel.component.kind` in the name of the metric vs as a label:
* The rule of thumb for having a label vs being in the name is if the sum across that label makes sense or not. I think actually it doesn't make that much sense.
* The mechanism of disabling (you can still disable the calculation, but not the recording since you don't know which labels will be recorded) and avoiding computation is at "metric" level. 

I think it may not be bad to have per component kind metrics. I would not necessary go to have per component kind and type level, but that can be another option.

Reply:
> The rule of thumb for having a label vs being in the name is if the sum across that label makes sense or not. I think actually it doesn't make that much sense.

I am on the fence about this. I agree the case is weak but aggregation along this dimension might at least give a sense of scale.

What I mean is that if you consider each component as doing some non-trivial amount of work for a amount of data, you can aggregate to understand relatively how much work the collector is doing internally. e.g. Say you are considering some changes to a configuration and compare v(n) to v(n+1). If you aggregate across all dimensions of "incoming" you may see that even with the same inputs the collector is having 10x items passed between components.

Either way, I'm not necessarily against having metrics per kind. Some consequences though:
- To have consistent attributes with other signals, we should still apply the `otel.signal` attribute, even though this is redundant with the metric name.
- It's more difficult for users to answer questions like, "which of the components is producing the most items" because they have to join across 3 metrics before sorting?

> The mechanism of disabling (you can still disable the calculation, but not the recording since you don't know which labels will be recorded) and avoiding computation is at "metric" level.

I'm not sure I understand. Are you pointing out that users would be forced to enable/disable the metric for all kinds, but maybe they should be able to control it for each kind independently? e.g. Only record items produced by receivers and consumed by exporters?

Reply:
Criteria discussedfrom 2024-10-21 meeting for continuing the conversation
- Looking at what we do in semantic conventions for similar cases (databases?)
- Thinking about discoverability, specifically in the context of additional, component kind specific metrics

Reply:
> Looking at what we do in semantic conventions for similar cases (databases?)

It's not clear to me that any existing semantic conventions are very relevant to this situation.

> Thinking about discoverability, specifically in the context of additional, component kind specific metrics

I think this was a good point. To elaborate on it, we may eventually define for example an additional metric that only applies to receivers. If so, the metric name will likely have a prefix such as `otelcol.receiver`, so it would be co-discoverable with `otelcol.receiver.produced.items` but not `otelcol.component.produced.items`.

Reply:
> To have consistent attributes with other signals, we should still apply the otel.signal attribute, even though this is redundant with the metric name.
I believe what bogdan is suggesting is putting the component kind in the metric name (i.e. `otelcol_receiver_produced_items`) which would avoid the redundancy here.

> It's more difficult for users to answer questions like, "which of the components is producing the most items" because they have to join across 3 metrics before sorting?
Maybe this is naive of me... but are there processors that increase the data produced? I guess this theoretically is possible, but I'm not aware of any processors that do this today. As a result, i think you would max across the receiver produced items.

I think my only qualm with this naming (including component kind in the measure name) comes down to user experience. If a user only wants minimal telemetry out of the collector, giving them as few signals (keeping `otelcol_component_produced_items`) as possible while still containing useful data would be advantageous.

Reply:
Based on the points made in this thread, I have updated the proposed metrics to include component kinds in their names.